step3 리뷰 요청.

hyeongwon-up opened the pull request.

hyeongwon-up committed:
docs(README.md): 3단계 로또 에 대한 기능 요구사항 정리.

hyeongwon-up committed:
feat: LottoGenerator 유틸리티 추가.
- 1-45 범위의 6자리 숫자를 가진 리스트를 반환하는 기능 추가 및 테스트.

hyeongwon-up committed:
feat: Lotto 도메인 추가.
- LottoGenerator 를 이용하여 디폴트 생성자를 이용할 경우 로또 생성.
- 직접 입력받은 인자로 로또 생성.
- 로또끼리 비교하여 카운트 해주는 기능 추가.

hyeongwon-up committed:
feat: LottoList 일급 컬렉션 도메인 추가.

hyeongwon-up committed:
feat: 사용자의 입력으로 부터 구매를 대행해주고, 로또를 발급해주는 Counter 모델 추가

hyeongwon-up committed:
feat: 지난 당첨 번호 입력 번호로 부터 문자열을 파싱하기 위한 SplitUtil 추가

hyeongwon-up committed:
feat: 지난 당첨 로또를 관리하는 모델 추가

hyeongwon-up committed:
feat: 로또의 등수와 당첨 금액을 관리하는 RankCode enum 추가

hyeongwon-up committed:
feat: 로또의 결과를 관리해줄 도메인 모델 추가

hyeongwon-up committed:
feat: 로또의 진행상황을 보여줄 InputView, OutputView 추가

hyeongwon-up committed:
feat: 로또의 진행상황을 할 수 있는 LottoController 와 LottoApplication 추가

geunhwanlee commented on src/main/java/lotto/README.md:
```diff
@@ -0,0 +1,60 @@
+# 3단계 - 로또(자동)
+
+## 기능 요구사항
```
기능목록 정리 👍🏻

geunhwanlee commented on src/main/java/lotto/code/RankCode.java:
```diff
@@ -0,0 +1,44 @@
+        .orElse(NOTHING);
+  }
+
+  public static Map<RankCode, Integer> generateRankCodeMap() {
```
사용하지 않는 메서드네요!
제거해도 될듯 합니다.

hyeongwon-up replied:
넵. 제거하였습니다. :D

geunhwanlee commented on src/main/java/lotto/code/RankCode.java:
```diff
@@ -0,0 +1,44 @@
+    return rankCodeMap;
+  }
+
+  public static int getMoney(RankCode rankCode) {
+    return RankCode.valueOf(rankCode.name()).money;
+  }
+
+  public static int containsCount(RankCode rankCode) {
+    return RankCode.valueOf(rankCode.name()).containsCount;
+  }
```
```
public int getMoney();
public int containsCount();
```
로 작성해도 될것 같은데
static 메서드로 작성한 이유가 무엇인지 궁금합니다!

hyeongwon-up replied:
작성을 할 때 돌이켜보면, enum 에 대해서 깊은 지식은 없고. 미리 정의해놓는 것 이라는 느낌의 클래스라는 생각을 가지고 있었는데요.
따라서 어디서든 바로바로 호출 할 수 있게 구현을 했었습니다.
다시 생각해보니 static 일 필요가 없군요. 관려해서 조금 더 공부를 하고 생각을 해보겠습니다.
해당 부분은 static 을 사용할 필요가 없어 제거하였습니다.

geunhwanlee commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,36 @@
+
+    List<Integer> targetLottoNumber = targetLotto.getNumbers();
+
+    for (Integer i : targetLottoNumber) {
+      if (this.numbers.contains(i)) {
+        count++;
+      }
+    }
```
indent가 2를 넘어가는데
어떻게 하면 indent를 줄일수 있을지 고민해보는것도 좋을것 같아요!

hyeongwon-up replied:
내부 if 문을 따로 메서드로 추출하여 리팩토링 해보았습니다.!

geunhwanlee commented on src/main/java/lotto/model/LottoList.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class LottoList {
```
일급컬렉션 👍🏻

geunhwanlee commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,36 @@
+
+  private final List<Integer> numbers;
+
+  public Lotto() {
+    numbers = LottoGenerator.generateLottoNumbers();
+  }
```
외부에서 `new Lotto();`로 생성할때 어떻게 동작하는지 드러나지 않는 것 같아요!

정적 팩토리 메서드를 사용해서 어떻게 동작할지 이름을 지정해줄수 있어요.
[정적 팩토리 메서드(Static Factory Method)는 왜 사용할까?](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)

hyeongwon-up replied:
오.. 좋은 방법인것 같아요. ㅎㅎ 감사합니다!!! 
수정하였습니다.

geunhwanlee commented on src/main/java/lotto/model/LottoResult.java:
```diff
@@ -0,0 +1,107 @@
+      RankCode rankCode = RankCode.getRankCode(matchingCount);
+
+      if (resultMap.containsKey(rankCode)) {
+        int value = resultMap.get(rankCode);
```
indent가 2를 넘어가네요!
어떻게 줄일수 있을지 고민해보면 좋을것 같아요.

hyeongwon-up replied:
넵 내부 메서드를 추출하여 수정하였습니다.

geunhwanlee commented on src/main/java/lotto/model/LottoResult.java:
```diff
@@ -0,0 +1,107 @@
+    RankCode rankCode = rankEntry.getKey();
+    StringBuilder stringBuilder = new StringBuilder();
+    return stringBuilder
+        .append(RankCode.containsCount(rankCode))
+        .append("개 일치 ")
+        .append("(")
+        .append(RankCode.getMoney(rankCode))
+        .append("원)- ")
+        .append(rankEntry.getValue())
+        .append("개").toString();
```
`String.format()`을 사용하면 더 간결하게 작성할수 있을것 같아요!

[참고 링크](https://velog.io/@yu-jin-song/JAVA-문자열-형식-지정)

hyeongwon-up replied:
훨씬 보기좋네요! 👍  수정하였습니다.

geunhwanlee commented on src/main/java/lotto/model/LottoResult.java:
```diff
@@ -0,0 +1,107 @@
+    return prizeMoney / (lottoList.getLottoList().size() * LOTTO_PRICE);
+  }
+
+  public String convertYieldToString() {
+    StringBuilder stringBuilder = new StringBuilder();
+    double yield = (Math.round((calculateYield() * MATH_ROUND_VALUE)) / MATH_ROUND_VALUE);
+    stringBuilder
+        .append("총 수익률은 ")
+        .append(yield)
+        .append("입니다.");
+    if (yield < COUNT_VALUE) {
+      stringBuilder.append("(기준이 1이기 떄문에 결과적으로 손해라는 의미임)");
+    }
+    return stringBuilder.toString();
+  }
```
LottoResult는 패키지 구조상 도메인 계층으로 보이는데요
위 로직은 뷰 계층의 관심사를 포함하고 있는것 같아요!
(콘솔에 어떤 글자를 보여줄지 만드는것)

제 생각에는 OutputView가 해당 역할을 해야할것 같은데 형원님의 생각은 어떠신가요?🤔

hyeongwon-up replied:
우선, 저도 미션을 진행하면서 고려했던 부분이였고 뷰 계층에 있어야 하는지에 대해서도 고민을 했었습니다.
제가 이렇게 생각했던 이유는 아래와 같은데요.

우선, LottoResult 라는 도메인이 로또의 결과에 대하여 제공해주는 도메인이라고 생각을하고 구현을 했습니다.
그리고 조건에 따라서  문구가 달라지는데요. 따라서 이러한 로직?또한 들어가므로 위 기능은 모델에 있는게 맞지않나? 라는 의구심에 위와 같이 진행을 하게 되었는데요.

어떻게 생각해보면 근환님 말씀처럼 뷰 계층에 관심사를 포함하고 있다고 보이기도 해요.

리뷰를 받으며 다시 생각해보면 뷰에 구현을 해야하나 싶기도 하고.. 또 그렇게하면 뷰가 모델 컴포넌트에 대해서 알고있어야 되는 부분이니 틀리지 않나? 라는 생각이 듭니다..

geunhwanlee replied:
LottoResult가 결과를 집계하여 메시지를 만드는 역할을 한다고 보면
지금처럼 뷰의 메시지를 만드는 로직이 있어도 괜찮을것 같아요!👍🏻

geunhwanlee commented on src/main/java/lotto/model/WinningLotto.java:
```diff
@@ -0,0 +1,29 @@
+  private Lotto generateWinningLotto(String winningLottoNumber) {
+    String[] splitNumbers = SplitUtil.splitInputNumbers(winningLottoNumber);
+
+    List<Integer> nums = Arrays.stream(splitNumbers).mapToInt(Integer::parseInt)
```
숫자가 아닌 값이 들어올 경우 NumberFormatException이 발생할것 같아요!

hyeongwon-up replied:
넵. 해당 부분에 대하여
1. 6자리 숫자 확인.
2. 숫자와 컴마 이외의 다른 입력
에 대한 validation 및 예외처리를 추가하였습니다.

geunhwanlee commented on src/main/java/lotto/util/SplitUtil.java:
```diff
@@ -0,0 +1,11 @@
+package lotto.util;
+
+public class SplitUtil {
```
static 메서드만을 제공하는 유틸리티 클래스는 private 생성자를 통해서 인스턴스화를 막을수 있어요!

[참고 링크](https://velog.io/@lychee/이펙티브-자바-아이템-4.-인스턴스화를-막으려거든-private-생성자를-사용하라)

hyeongwon-up replied:
한번도 생각하지 못한 부분이였는데, 감사합니다.
수정하였습니다.

geunhwanlee commented on src/main/java/lotto/util/LottoGenerator.java:
```diff
@@ -0,0 +1,30 @@
+  private static final int MIN_LOTTO_NUMBER = 1;
+  private static final int MAX_LOTTO_NUMBER = 45;
+  private static final int MIN_LOTTO_INDEX = 0;
+  private static final int MAX_LOTTO_INDEX = 6;
```
```suggestion
  private static final int LOTTO_SIZE = 6;
```
이렇게 하는게 의미가 명확할것 같은데 어떤가요?🤔

hyeongwon-up replied:
좋은 것 같은데 . 
한가지 쫌 찜찜한 것이 
`  private static final int MIN_LOTTO_INDEX = 0;
`
는 다른 사람이 보기에 헷갈릴 수도 있겠단 생각이 드네요.. 그래서 통일은 하긴 했었는데..  어떻게 생각하시나요?

우선 수정하였습니다.

geunhwanlee replied:
MIN_LOTTO_INDEX는 ZERO_INDEX 정도로 해볼까요?🤔

geunhwanlee commented on src/test/java/lotto/model/CounterTest.java:
```diff
@@ -0,0 +1,35 @@
+    //when
+    Counter counter = new Counter(inputMoney);
+    //then
+    assertThat(counter.getLottoList().getLottoList().size()).isEqualTo(14);
```
```suggestion
    assertThat(counter.getLottoList().getLottoList()).hasSize(14);
```
이렇게 작성하면 더 가독성이 좋을것 같아요!

hyeongwon-up replied:
넵 ㅎㅎ 수정하였습니다.

geunhwanlee reviewed (changes requested):
3단계 미션 잘해주셨습니다.👍🏻

몇가지 코멘트를 남겼는데요
꼭 정답이 있는건 아니니 형원님의 생각을 말씀해주셔도 됩니다!

그럼 화이팅입니다!🔥

hyeongwon-up committed:
refactor(RankCode.java): 사용하지 않는 메서드 삭제.

hyeongwon-up committed:
refactor(RankCode.java, LottoResult): static 제거.

hyeongwon-up committed:
refactor(Lotto.java, LottoResult.java):
- Lotto의 메서드명 변경 countMatchingNumber -> getMatchingCount
- getMatchingCount 내부 리팩토링

hyeongwon-up committed:
refactor(Lotto.java): 오버로딩한 생성자를 사용하지 않고 정적 팩토리 메서드를 이용하도록 수정.

hyeongwon-up committed:
refactor(LottoResult.java): calculateLottoResult() 내부 리팩토링.

hyeongwon-up committed:
refactor(LottoResult.java): String.format 을 사용

hyeongwon-up committed:
refactor(SplitUtil.java, LottoGenerator.java): 유틸리티 클래스 인스턴스화를 하지 못하도록 수정

hyeongwon-up committed:
refactor(LottoResult.java): 잘못된 calculateLottoMatch() 로직 수정

hyeongwon-up committed:
refactor(WinningLotto.java): 지난 당청 번호 입력에 대한 벨리데이션 추가 및 예외 처리.

hyeongwon-up committed:
test(CounterTest.java): 테스트 확인 간소화

hyeongwon-up committed:
refactor(LottoGenerator.java): static 상수 이름 수정.

geunhwanlee commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,43 @@
+
+    List<Integer> targetLottoNumber = targetLotto.getNumbers();
+
+    for (int number : targetLottoNumber) {
+      count = this.countContainNumber(number, count);
+    }
+
+    return count;
```
```suggestion
    return (int) targetLottoNumber.stream()
            .filter(numbers::contains)
            .count();
```
이 방법은 어떤가요?

geunhwanlee commented on src/main/java/lotto/model/WinningLotto.java:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  private void validWinningLottoNumber(String winningLottoNumber) {
+    if(!winningLottoNumber.matches(NUMBER_REGEX)) {
```
String.matches는 실행시점마다 Pattern 객체를 생성하게 되어있어요.
참고만 해주세요.😀

[Pattern.matcher VS String.match](https://velog.io/@edwin/JAVA-정규식-성능에-대해-Pattern.matcher-VS-String.matches)

geunhwanlee reviewed (approved):
리뷰사항 잘 반영해 주셨어요👍🏻

이번 단계는 여기서 머지하도록 할게요.

소소한 코멘트 남겼으니 확인해보시고
다음 단계도 화이팅하세요!🔥
